src/psyclone/psyir/nodes/call.py

```diff
@@ -592,4 +634,189 @@ def _location_txt(node):
+        :rtype: None|List[int]
+        """
+
+        # Create a copy of the list of actual arguments to the routine.
+        # Once an argument has been successfully matched, set it to 'None'
+        routine_argument_list: List[DataSymbol] = (
+            routine.symbol_table.argument_list[:]
+        )
+
+        if len(self.arguments) > len(routine.symbol_table.argument_list):
+            raise CallMatchingArgumentsNotFound(
+                f"More arguments in call ('{self.debug_string()}')"
+                f" than callee (routine '{routine.name}')"
+            )
+
+        # Iterate over all arguments to the call
+        ret_arg_idx_list = []
+        for call_arg_idx, call_arg in enumerate(self.arguments):
+            call_arg_idx: int
+            call_arg: DataSymbol
+
+            # If the associated name is None, it's a positional argument
+            # => Just return the index if the types match
+            if self.argument_names[call_arg_idx] is None:
+                routine_arg = routine_argument_list[call_arg_idx]
+                routine_arg: DataSymbol
+
+                # Do the types of arguments match?
+                #
+                # TODO #759: If optional is used, it's an unsupported Fortran
+                # type and we need to use the following workaround
+                # Once this issue is resolved, simply remove this if branch.
+                # Optional arguments are processed further down.
+                if not isinstance(
+                    routine_arg.datatype, UnsupportedFortranType
+                ):
+                    if call_arg.datatype != routine_arg.datatype:
+                        raise CallMatchingArgumentsNotFound(
+                            f"Argument type mismatch of call argument "
+                            f"'{call_arg}' and routine argument "
+                            f"'{routine_arg}'"
+                        )
+
+                ret_arg_idx_list.append(call_arg_idx)
+                routine_argument_list[call_arg_idx] = None
+                continue
+
+            #
+            # Next, we handle all named arguments
+            #
+            arg_name = self.argument_names[call_arg_idx]
+            routine_arg_idx = None
+
+            for routine_arg_idx, routine_arg in enumerate(
+                routine_argument_list
+            ):
+                routine_arg: DataSymbol
+
+                # Check if argument was already processed
+                if routine_arg is None:
+                    continue
+
+                if arg_name == routine_arg.name:
+                    # TODO #759: If optional is used, it's an unsupported
+                    # Fortran type and we need to use the following workaround
+                    # Once this issue is resolved, simply remove this if
+                    # branch.
+                    # Optional arguments are processed further down.
+                    if isinstance(
```

Comment:
Same here, you should be able to use `partial_datatype` and drop the check for 'OPTIONAL' later.

Reply:
OK. Replaced by `_check_argument_type_matches` method.